src/httpauth/digest.c

```diff
@@ -410,3 +413,190 @@ int httpauth_digest_response_encode(const struct httpauth_digest_resp *resp,
+	if (!hash) {
+		err = ENOMEM;
+		goto out;
+	}
+
+	if (str_isset(secret))
+		err = mbuf_printf(mb, "%lu:%s:%s",
+			(long unsigned)ts, etag, secret);
+	else
+		err = mbuf_printf(mb, "%lu:%s", (long unsigned)ts, etag);
+
+	if (err)
+		goto out;
+
+	sha256(mb->buf, mb->end, hash);
+	mbuf_rewind(mb);
+
+	err = mbuf_printf(mb, "%w%016lx", hash, hashlen, (long unsigned)ts);
+	if (err)
+		goto out;
+
+	mbuf_set_pos(mb, 0);
+	err = mbuf_strdup(mb, &nonce, mbuf_get_left(mb));
+
+out:
+	if (err)
+		mem_deref(nonce);
+	else
+		*pnonce = nonce;
+
+	mem_deref(mb);
+	mem_deref(hash);
+
+	return err;
+}
+
+
+/**
+ * Prints / encodes an HTTP digest request challenge
+ *
+ * @param pf  Re_printf object
+ * @param req Request to print
+ *
+ * @return 0 if success, otherwise errorcode
+ */
+int httpauth_digest_chall_req_print(struct re_printf *pf,
+	const struct httpauth_digest_chall_req *req)
+{
+	int err = 0;
+
+	if (!req)
+		return EINVAL;
+
+	/* historical reason quoted strings:   */
+	/*   realm, domain, nonce, opaque, qop */
+	/* historical reason unquoted strings: */
+	/*   stale, algorithm                  */
+	err = re_hprintf(pf, "Digest realm=\"%s\", "
+		"qop=\"%s\", nonce=\"%s\", algorithm=%s",
+		req->realm, req->qop, req->nonce, req->algorithm);
+
+	if (str_isset(req->opaque))
+		err |= re_hprintf(pf, ", opaque=\"%s\"", req->opaque);
+	if (str_isset(req->domain))
+		err |= re_hprintf(pf, ", domain=\"%s\"", req->domain);
+	if (req->stale)
+		err |= re_hprintf(pf, ", stale=true");
+	if (str_isset(req->charset))
+		err |= re_hprintf(pf, ", charset=\"%s\"", req->charset);
+	if (req->userhash)
+		err |= re_hprintf(pf, ", userhash=true");
+
+	return err;
+}
+
+
+/**
+ * Minimal requirement digest authentication request
+ *
+ * @param preq  Httpauth_digest_chall_req object ptr
+ * @param realm Realm
+ * @param etag  Changing data for nonce creation
+ *              (HTTP ETag header / SIP msg src address)
+ * @param qop   Quality of protection
+ *
+ * @return 0 if success, otherwise errorcode
+ */
+int httpauth_digest_min_chall_request(struct httpauth_digest_chall_req **preq,
+	const char *realm, const char *etag, const char *qop)
+{
+	return httpauth_digest_chall_request(preq, realm, NULL, etag, NULL,
+		false, NULL, qop, NULL, false);
+}
+
+
+/**
+ * Create a digest authentication request
+ *
+ * @param preq      Httpauth_digest_chall_req object ptr
+ * @param realm     Realm
+ * @param domain    Domain (not used in SIP)
+ * @param etag      Changing data for nonce creation
+ *                  (HTTP ETag header / SIP msg src address)
+ * @param opaque    Opaque
+ * @param stale     Stale
+ * @param algo      Supported algorithm (MD5, SHA1, SHA256 and sess versions)
+ * @param qop       Quality of protection
+ * @param charset   Character set used (not used in SIP)
+ * @param userhash  Userhash support (not used in SIP)
+ *
+ * @return 0 if success, otherwise errorcode
+ */
+int httpauth_digest_chall_request(struct httpauth_digest_chall_req **preq,
```

Comment:
I suggest to call this function:
`httpauth_digest_chall_request_full()`
and the short form
`httpauth_digest_chall_request()`